### DISCUSSION TITLE
Is it possible we support stream?

### DISCUSSION THREAD

**Discussion by addlistener** [trpc/trpc#3169] [Ideas]
https://github.com/trpc/trpc/search?q=stream

It's really useful and we could provide an option to specify the implementation.

**Comment by KATT** [comment]
Can you elaborate what you mean?

**Comment by BishalN** [comment]
Can we stream partial responses like in this case I am working with openai sdk and it supports this streaming 
```ts
const res = await openai.createCompletion({
  model: "text-davinci-002",
  prompt: "Say this is a test",
  max_tokens: 6,
  temperature: 0,
  stream: true,
}, { responseType: 'stream' });

res.on('data', console.log)
```

How can we achieve this in trpc ?

**Comment by masnwilliams** [comment]
Any updates on this?

**Comment by addlistener** [comment]
I've already tackled it by using subscriptions https://trpc.io/docs/subscriptions. It's just what I need @johnmasonwilliams

**Comment by helmturner** [comment]
Subscriptions are still not ideal, as that rules out deployment on Vercel. SSE can be implemented on Vercel using the Edge runtime, but the edge runtime seems to break the tRPC api handler.

**Comment by jamesmfriedman** [comment]
Thinking about this too. Dealing with some large JSON files I'd like to stream through the backend. Haven't dug into this codebase but at face value it seems like something that could be supported by a flag like proposed and the client would handle reading the stream and returning the fully resolved response.

**Reply by helmturner** [reply]
I'm working on an implementation now. In the end, it should support both streaming via SSE with 'Content-Type': 'text/event-stream', and via 'Transfer-Encoding': 'chucked' with 'Content-Type': 'application/json'.

Just setting the 'Accept' header on the client request is all it would take to switch between them.

**Comment by joseph-mccombs** [comment]
is there any update on this?

**Comment by helmturner** [comment]
Yo! Sorry for the radio silence - I had a priority project come up that I had to focus on, but this is now back on my plate! 🚧👷‍♂️

**Reply by cyrus-za** [reply]
Is there some place we can view your work in progress code?

**Comment by yaminagile** [comment]
Is there any updated on this?

**Comment by tgdn** [comment]
Is there any update on this?

**Comment by helmturner** [comment]
@tgdn @yaminagile @joseph-mccombs @cyrus-za Quick update: Seeing good progress; Most tests passing currently. A few semantics to work out still. If you'd like to help get it over the hump check out [this draft pr](https://github.com/trpc/trpc/pull/4530) (where I've added more information) and see if you can get any of the last few tests passing 😉

**Reply by mattgraphlan** [reply]
And https://github.com/trpc/trpc/pull/4911

**Comment by aleclarson** [comment]
Hey everyone! I'm the creator of [Alien RPC](https://github.com/alloc/alien-rpc), which has JSON streaming powered by JavaScript “async generators” and the [JSON Text Sequence](https://www.rfc-editor.org/rfc/rfc7464.html) proposed standard.

For an example route, check out [this test case](https://github.com/alloc/alien-rpc/blob/4d3c772c7e2cf18812e51bf7b557b5afcabe5db5/test/generator/__fixtures__/pagination/routes.ts).

**Reply by KATT** [reply]
We support streaming too through [`httpBatchStreamLink`](https://trpc.io/docs/client/links/httpBatchStreamLink).

**Reply by FunctionDJ** [reply]
@KATT ~~As far as i can tell, `httpBatchStreamLink` is for streaming/batching many requests, but the reply to each request will only start to send once it's completely done - you can't reply with e.g. a [`ReadableStream`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream). That means using tRPC to stream large text as a reply to a single query is not possible.
Please correct me if i'm wrong, but that's what OP is asking for (and me too), not batching.~~

Edit: I was wrong, tRPC supports streaming async generators, see below comments.

**Reply by KATT** [reply]
I'm not sure what you mean - your query can be an async generator function  https://trpc.io/docs/client/links/httpBatchStreamLink#generators

**Reply by FunctionDJ** [reply]
@KATT Pardon, i missed that part of the docs and assumed that this wasn't supported yet.
I've just tried it out and it works nicely, at least for strings.
Can this also work for data that isn't strings, like binary? Without using base64?

**Reply by KATT** [reply]
Only strings at the moment / foreseeable future, that's all we have need ourselves. We'd love to support more content types (and refactor our response handling) but we have limited time

If you'd be down to contribute, feel free to hit us up on discord.

**Reply by brunolm** [reply]
Where can I find an example of this? I created the async generator function in a route, but how do I consume it? `trpc.thing.endpoint.fetch` or what?

**Comment by devguilhermy** [comment]
Updates?

**Reply by BeBoRE** [reply]
https://trpc.io/docs/client/react/useQuery#streaming

**Reply by BeBoRE** [reply]
https://trpc.io/docs/client/links/httpBatchStreamLink#generators
